crates/sparrow-plan-execution/src/write_pipeline.rs

```diff
@@ -0,0 +1,214 @@
+use std::collections::VecDeque;
+use std::sync::atomic::{AtomicBool, Ordering};
+use std::sync::Arc;
+
+use arrow_schema::SchemaRef;
+use error_stack::ResultExt;
+use parking_lot::Mutex;
+use sparrow_batch::Batch;
+use sparrow_interfaces::destination::{Destination, Writer};
+use sparrow_interfaces::types::{Partition, Partitioned};
+use sparrow_scheduler::{Pipeline, PipelineError, Scheduler, TaskRef};
+
+/// Pipeline for writing to a destination.
+#[derive(Debug)]
+pub struct WritePipeline {
+    destination: Arc<dyn Destination>,
+    #[allow(unused)]
+    schema: SchemaRef,
+    partitions: Partitioned<WritePartition>,
+}
+
+#[derive(Debug)]
+struct WritePartition {
+    task: TaskRef,
+    is_closed: AtomicBool,
+    inputs: Mutex<VecDeque<Batch>>,
+    writer: WriterState,
+}
+
+impl WritePartition {
+    /// Close the input. Returns true if the input buffer is empty.
+    fn close(&self) -> bool {
+        self.is_closed.store(true, Ordering::Release);
+        self.inputs.lock().is_empty()
+    }
+
+    fn is_closed(&self) -> bool {
+        self.is_closed.load(Ordering::Acquire)
+    }
+
+    fn add_input(&self, batch: Batch) {
+        self.inputs.lock().push_back(batch);
+    }
+
+    fn pop_input(&self) -> Option<Batch> {
+        self.inputs.lock().pop_front()
+    }
+}
+
+#[derive(Debug)]
+enum WriterState {
+    NotStarted,
+    Writing(Mutex<Box<dyn Writer>>),
```

Comment:
I think you can eliminate the box since the `Mutex` already implies it's on the heap, which eliminates an indirection:

```Mutex<dyn Writer>```

Reply:
Hm, I thought that a `dyn` needed to be boxed because the implementations are not statically sized:
```
the size for values of type `(dyn Writer + 'static)` cannot be known at compilation time
within `parking_lot::lock_api::Mutex<parking_lot::RawMutex, (dyn Writer + 'static)>`, the trait `Sized` is not implemented for `(dyn Writer + 'static)`
no field of an enum variant may have a dynamically sized type
change the field's type to have a statically known size
```

Is there anything I'm not understanding there?

Reply:
Hmm. Maybe I missed it. Things need to be on the heap (boxed, arc, etc.) in order to be `dyn`. I was thinking that `Mutex` would support that, and thought I had tested that in Rust Playground, but apparently my test wasn't deep enough. It looks like we do need it in the box.